File: wrench-prefs/src/main/java/com/izettle/wrench/preferences/WrenchPreferences.java
package com.izettle.wrench.preferences;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;

import com.izettle.wrench.core.Bolt;
import com.izettle.wrench.core.Nut;
import com.izettle.wrench.core.WrenchProviderContract;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class WrenchPreferences {
    private final ContentResolver contentResolver;

    public WrenchPreferences(@NonNull Context context) {
        this.contentResolver = context.getContentResolver();
    }

    private static void insertNut(@NonNull ContentResolver contentResolver, @NonNull Nut nut) {
        contentResolver.insert(WrenchProviderContract.nutUri(), nut.toContentValues());
    }

    @Nullable
    private static Bolt getBolt(@NonNull ContentResolver contentResolver, @NonNull @Bolt.BoltType String boltType, @NonNull String key) {
        Cursor cursor = null;
        try {
            cursor = contentResolver.query(WrenchProviderContract.boltUri(key),
                    null,
                    null,
                    null,
                    null);
            if (cursor == null) {
                return null;
            }

            if (cursor.moveToFirst()) {
                return Bolt.fromCursor(cursor);
            }


        } finally {
            if (cursor != null && !cursor.isClosed()) {
                cursor.close();
            }
        }
        return new Bolt(0L, boltType, key, null);
    }

    private static Uri insertBolt(@NonNull ContentResolver contentResolver, @NonNull Bolt bolt) {
        return contentResolver.insert(WrenchProviderContract.boltUri(), bolt.toContentValues());
    }

    @NonNull
    public <T extends Enum<T>> T getEnum(@NonNull String key, @NonNull Class<T> type, @NonNull T defValue) {
        Bolt bolt = getBolt(contentResolver, Bolt.TYPE.ENUM, key);
        if (bolt == null) {
            return defValue;
        }

        if (bolt.getId() == 0) {
            bolt = bolt.copy(bolt.getId(), bolt.getType(), bolt.getKey(), defValue.toString());
            Uri uri = insertBolt(contentResolver, bolt);
            bolt.setId(Long.parseLong(uri.getLastPathSegment()));

            for (T enumConstant : type.getEnumConstants()) {
                insertNut(contentResolver, new Nut(bolt.getId(), enumConstant.toString()));
            }
        }

        if (bolt.getValue() == null) {
            throw new IllegalStateException("bolt value cannot be null for enum");
        }
        return Enum.valueOf(type, bolt.getValue());
    }

    @Nullable
    public String getString(@NonNull String key, @Nullable String defValue) {

        Bolt bolt = getBolt(contentResolver, Bolt.TYPE.STRING, key);
        if (bolt == null) {
            return defValue;
        }

        if (bolt.getId() == 0) {
            bolt = bolt.copy(bolt.getId(), bolt.getType(), bolt.getKey(), defValue);
            Uri uri = insertBolt(contentResolver, bolt);
            bolt.setId(Long.parseLong(uri.getLastPathSegment()));
        }

        return bolt.getValue();
    }

    public boolean getBoolean(@NonNull String key, boolean defValue) {
        Bolt bolt = getBolt(contentResolver, Bolt.TYPE.BOOLEAN, key);
        if (bolt == null) {
            return defValue;
        }

        if (bolt.getId() == 0) {
            bolt = bolt.copy(bolt.getId(), bolt.getType(), bolt.getKey(), String.valueOf(defValue));
            Uri uri = insertBolt(contentResolver, bolt);
            bolt.setId(Long.parseLong(uri.getLastPathSegment()));
        }

        return Boolean.valueOf(bolt.getValue());
    }

    public int getInt(@NonNull String key, int defValue) {
        Bolt bolt = getBolt(contentResolver, Bolt.TYPE.INTEGER, key);
        if (bolt == null) {
            return defValue;
        }

        if (bolt.getId() == 0) {
            bolt = bolt.copy(bolt.getId(), bolt.getType(), bolt.getKey(), String.valueOf(defValue));
            Uri uri = insertBolt(contentResolver, bolt);
            bolt.setId(Long.parseLong(uri.getLastPathSegment()));
        }

        return Integer.valueOf(bolt.getValue());
    }
}

File: wrench-prefs/src/main/java/com/izettle/wrench/preferences/WrenchPreferences.kt
package com.izettle.wrench.preferences

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import com.izettle.wrench.core.Bolt
import com.izettle.wrench.core.Nut
import com.izettle.wrench.core.WrenchProviderContract

open class WrenchPreferences(context: Context) {
    private val contentResolver: ContentResolver = context.contentResolver

    private fun insertNut(contentResolver: ContentResolver, nut: Nut) {
        contentResolver.insert(WrenchProviderContract.nutUri(), nut.toContentValues())
    }

    private fun getBolt(contentResolver: ContentResolver, @Bolt.BoltType boltType: String, key: String): Bolt? {
        val cursor = contentResolver.query(WrenchProviderContract.boltUri(key), null, null, null, null)
        cursor.use {
            if (cursor == null) {
                return null
            }

            if (cursor.moveToFirst()) {
                return Bolt.fromCursor(cursor)
            }
        }

        return Bolt(type = boltType, key = key)
    }

    private fun insertBolt(contentResolver: ContentResolver, bolt: Bolt): Uri? {
        return contentResolver.insert(WrenchProviderContract.boltUri(), bolt.toContentValues())
    }

    fun <T : Enum<T>> getEnum(key: String, type: Class<T>, defValue: T): T {
        var bolt = getBolt(contentResolver, Bolt.TYPE.ENUM, key) ?: return defValue

        if (bolt.id == 0L) {
            bolt = bolt.copy(key = key, type = Bolt.TYPE.ENUM, value = defValue.toString())
            val uri = insertBolt(contentResolver, bolt)
            bolt.id = uri!!.lastPathSegment.toLong()

            for (enumConstant in type.enumConstants) {
                insertNut(contentResolver, Nut(bolt.id, enumConstant.toString()))
            }
        }

        return java.lang.Enum.valueOf<T>(type, bolt.value!!)
    }

    fun getString(key: String, defValue: String?): String? {

        var bolt = getBolt(contentResolver, Bolt.TYPE.STRING, key) ?: return defValue

        if (bolt.id == 0L) {
            bolt = bolt.copy(key = key, type = Bolt.TYPE.STRING, value = defValue)
            insertBolt(contentResolver, bolt)
        }

        return bolt.value
    }

    fun getBoolean(key: String, defValue: Boolean): Boolean {
        var bolt = getBolt(contentResolver, Bolt.TYPE.BOOLEAN, key) ?: return defValue

        if (bolt.id == 0L) {
            bolt = bolt.copy(key = key, type = Bolt.TYPE.BOOLEAN, value = defValue.toString())
            insertBolt(contentResolver, bolt)
        }

        return java.lang.Boolean.valueOf(bolt.value)!!
    }

    fun getInt(key: String, defValue: Int): Int {
        var bolt = getBolt(contentResolver, Bolt.TYPE.INTEGER, key) ?: return defValue

        if (bolt.id == 0L) {
            bolt = bolt.copy(key = key, type = Bolt.TYPE.INTEGER, value = defValue.toString())
            insertBolt(contentResolver, bolt)
        }

        return Integer.valueOf(bolt.value)!!
    }
}
